In [1]:
# imports
import json
import pandas as pd

# EDA - User Data

In [5]:
# import json data
users = []
for line in open('users.json', 'r'):
    users.append(json.loads(line))
    
# put data into dataframe
users = pd.DataFrame(users)

In [6]:
users.head()

_id  active               createdDate  \
0  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   

                  lastLogin      role signUpSource state  
0  {'$date': 1609687537858}  consumer        Email    WI  
1  {'$date': 1609687537858}  consumer        Email    WI  
2  {'$date': 1609687537858}  consumer        Email    WI  
3  {'$date': 1609687530597}  consumer        Email    WI  
4  {'$date': 1609687537858}  consumer        Email    WI

In [7]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   _id           495 non-null    object
 1   active        495 non-null    bool  
 2   createdDate   495 non-null    object
 3   lastLogin     433 non-null    object
 4   role          495 non-null    object
 5   signUpSource  447 non-null    object
 6   state         439 non-null    object
dtypes: bool(1), object(6)
memory usage: 23.8+ KB


### Takeaways

    1. There are 495 user IDs provided. Are all unique or any duplicates?
    2. There is an extra column (signUpSource) that was not provided in the data schema.

## EDA of _id column

In [8]:
users.columns

Index(['_id', 'active', 'createdDate', 'lastLogin', 'role', 'signUpSource',
       'state'],
      dtype='object')

In [9]:
# Return the rows with duplicate user IDs
dupl_users = users.loc[users['_id'].duplicated(),:]
dupl_users

_id  active               createdDate  \
1    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
4    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
5    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
8    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
..                                    ...     ...                       ...   
490  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
491  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
492  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
493  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
494  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   

                    lastLogin         role signUpSource state  
1    {'$date': 1609687537858}     consumer        Email    WI  
2    {'$date': 1609687537858}     consumer        Email    WI  
4    {'$date': 1609687537858}     consumer        Email    WI  
5    {'$date': 1609687537858}     consumer        Email    WI  
8    {'$date': 1609687537858}     consumer        Email    WI  
..                        ...          ...          ...   ...  
490  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
491  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
492  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
493  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
494  {'$date': 1614963143204}  fetch-staff          NaN   NaN  

[283 rows x 7 columns]

### Takeaways
1. There are duplicate user IDs in the users data. _id is the primary key and values in this column should be unique. 
2. Looks like the duplicate user IDs have the same creation date. Maybe they were all created on similar dates.
3. Some of the duplicate user IDs were created by fetch-staff. Maybe test accounts?